wasm/src/main/java/com/dylibso/chicory/wasm/types/OpCode.java

```diff
@@ -0,0 +1,650 @@
+    // i64.trunc_f64_s
+    I64_TRUNC_F64_S(0xB0),
+    // i64.trunc_f64_u
+    I64_TRUNC_F64_U(0xB1),
+    // f32.convert_i32_s
+    F32_CONVERT_I32_S(0xB2),
+    // f32.convert_i32_u
+    F32_CONVERT_I32_U(0xB3),
+    // f32.convert_i64_s
+    F32_CONVERT_I64_S(0xB4),
+    // f32.convert_i64_u
+    F32_CONVERT_I64_U(0xB5),
+    // f32.demote_f64
+    F32_DEMOTE_F64(0xB6),
+    // f64.convert_i32_s
+    F64_CONVERT_I32_S(0xB7),
+    // f64.convert_i32_u
+    F64_CONVERT_I32_U(0xB8),
+    // f64.convert_i64_s
+    F64_CONVERT_I64_S(0xB9),
+    // f64.convert_i64_u
+    F64_CONVERT_I64_U(0xBA),
+    // f64.promote_f32
+    F64_PROMOTE_F32(0xBB),
+    // i32.reinterpret_f32
+    I32_REINTERPRET_F32(0xBC),
+    // i64.reinterpret_f64
+    I64_REINTERPRET_F64(0xBD),
+    // f32.reinterpret_i32
+    F32_REINTERPRET_I32(0xBE),
+    // f64.reinterpret_i64
+    F64_REINTERPRET_I64(0xBF),
+    // i32.extend_8_s
+    I32_EXTEND_8_S(0xC0),
+    // i32.extend_16_s
+    I32_EXTEND_16_S(0xC1),
+    // i64.extend_8_s
+    I64_EXTEND_8_S(0xC2),
+    // i64.extend_16_s
+    I64_EXTEND_16_S(0xC3),
+    // i64.extend_32_s
+    I64_EXTEND_32_S(0xC4),
+    // ref.null <varuint>
+    REF_NULL(0xD0),
+    // ref.is_null
+    REF_IS_NULL(0xD1),
+    // ref.func <varuint>
+    REF_FUNC(0xD2),
+    // i32.trunc_sat_f32_s
+    I32_TRUNC_SAT_F32_S(0xFC00),
+    // i32.trunc_sat_f32_u
+    I32_TRUNC_SAT_F32_U(0xFC01),
+    // i32.trunc_sat_f64_s
+    I32_TRUNC_SAT_F64_S(0xFC02),
+    // i32.trunc_sat_f64_u
+    I32_TRUNC_SAT_F64_U(0xFC03),
+    // i64.trunc_sat_f32_s
+    I64_TRUNC_SAT_F32_S(0xFC04),
+    // i64.trunc_sat_f32_u
+    I64_TRUNC_SAT_F32_U(0xFC05),
+    // i64.trunc_sat_f64_s
+    I64_TRUNC_SAT_F64_S(0xFC06),
+    // i64.trunc_sat_f64_u
+    I64_TRUNC_SAT_F64_U(0xFC07),
+    // memory.init <varuint> <varuint>
+    MEMORY_INIT(0xFC08),
+    // data.drop <varuint>
+    DATA_DROP(0xFC09),
+    // memory.copy <varuint> <varuint>
+    MEMORY_COPY(0xFC0A),
+    // memory.fill <varuint>
+    MEMORY_FILL(0xFC0B),
+    // table.init <varuint> <varuint>
+    TABLE_INIT(0xFC0C),
+    // elem.drop <varuint>
+    ELEM_DROP(0xFC0D),
+    // table.copy <varuint> <varuint>
+    TABLE_COPY(0xFC0E),
+    // table.grow <varuint>
+    TABLE_GROW(0xFC0F),
+    // table.size <varuint>
+    TABLE_SIZE(0xFC10),
+    // table.fill <varuint>
+    TABLE_FILL(0xFC11);
+
+    private final int opcode;
+
+    OpCode(int opcode) {
+        this.opcode = opcode;
+    }
+
+    public int opcode() {
+        return opcode;
+    }
+
+    private static final Map<Integer, OpCode> byOpCode = new HashMap<>();
+
+    public static OpCode byOpCode(int opcode) {
+        return byOpCode.get(opcode);
+    }
+
+    private static final Map<OpCode, WasmEncoding[]> signature = new HashMap<>();
+
+    public static WasmEncoding[] getSignature(OpCode o) {
+        return signature.get(o);
+    }
+
+    static {
+        for (OpCode e : OpCode.values()) {
+            byOpCode.put(e.opcode(), e);
```

Comment:
Might want a validation here
```java
if (byOpCode.put(e.opcode(), e) != null) {
    throw new AssertionError("duplicate opcode: " + e.opcode());
}

Reply:
Moved the Map to be an array, I see the value of the check, but is, ultimately, only for us developers of the engine.
I left it out for now.

Reply:
Agreed, and you would find the error immediately anyway, so it's not needed.